### ISSUE TITLE
onerng - verify (3.1)

### ISSUE THREAD

**Issue by noci2012** [OneRNG/firmware#1]
The verify tools fails with the message 
x = ord(c[0]) 
that the argument should be s 1 character string and .not an int.
removing ord() helps here.
After that the firmware fails the validity test (preventing startup in default config).

Manual startup of rngd does work. just like disabling verify and then starting the onerng.

Also if onerng doesn't start the kill function fails with a grep usage output, so it most probably misses some info.

**Comment by MoonbaseOtago** [comment]
On Sat, 11 Apr 2015 05:46:12 noci2012 wrote:

> The verify tools fails with the message 
> x = ord(c[0]) 
> that the argument should be s 1 character string and .not an int.
> removing ord() helps here.
> After that the firmware fails the validity test (preventing startup in
> default config).
> 
> Manual startup of rngd does work. just like disabling verify and then
> starting the onerng.
> 
> Also if onerng doesn't start the kill function fails with a grep usage
> output, so it most probably misses some info.

sounds like you're not running python3, we must have a dependency wrong - 
which system are you running on?

```
Paul
```

**Comment by noci2012** [comment]
gentoo, python3 tells me it's version is:

$ python3 --version
Python 3.4.1

and default python is:
$ python --version
Python 3.4.1

**Comment by noci2012** [comment]
was looking into the grep error is saw: and it is caused
    t1="`ls -lt /sys/class/tty/ttyACM* | grep $DEVPATH`"
where DEVPATH is non existent. Now i am not running systemd but using eudev instead.
hence the usage message on grep i saw.

btw i am not sure how this would be linked to a program ls -lt /sys/class/tty/ttyACM*
lrwxrwxrwx 1 root root 0 Apr 11 01:37 /sys/class/tty/ttyACM0 -> ../../devices/pci0000:00/0000:00:12.0/usb5/5-4/5-4:1.0/tty/ttyACM0

t2 would become ttyACM0..., what if there is more than one ttyACM0 (f.e. for a GSM dialer connected, or any USB/serial for that matter).
(or is this code that doesn;t work from the command line..., and as such just an artefact?).

**Comment by MoonbaseOtago** [comment]
On Sun, 12 Apr 2015 13:02:36 noci2012 wrote:

> was looking into the grep error is saw: and it is caused
>     t1="`ls -lt /sys/class/tty/ttyACM* | grep $DEVPATH`"
> where DEVPATH is non existent. Now i am not running systemd but using eudev
> instead. hence the usage message on grep i saw.
> 
> btw i am not sure how this would be linked to a program ls -lt
> /sys/class/tty/ttyACM\* lrwxrwxrwx 1 root root 0 Apr 11 01:37
> /sys/class/tty/ttyACM0 ->
> ../../devices/pci0000:00/0000:00:12.0/usb5/5-4/5-4:1.0/tty/ttyACM0
> 
> t2 would become ttyACM0..., what if there is more than one ttyACM0 (f.e. for
> a GSM dialer connected, or any USB/serial for that matter). (or is this
> code that doesn;t work from the command line..., and as such just an
> artefact?).

ah - so what's going on there is that there's a bug in some UDEV 
implementations where it passes the wrong device name into the removal script 
- what the grep is searching for is a link that contains the bogus parameter - 
  looks like Gentoo's UDEV is broken in a different way from Debian ....

(you wont kill another ttyACMx device because the $DEVPATH would not match)

It's unlikely this is going to work from the command line because the issue is 
that UDEV is sort of half run when it's doing this, the structures in /sys for 
the device (ttyACM0 or whatever) are mostly gone - there's one dangling 
symbolic link we can use to make the connection between the bad parameter and 
the actual device name 

(if it helps redhat seem to have fixed this UDEV bug)

I was a Gentoo fan for a long time, I kind of shudder at the effort to

**Comment by noci2012** [comment]
this is DEVPATH when invoked from udev.
Mon Apr 13 01:57:54 CEST 2015 DEVPATH=/devices/pci0000:00/0000:00:12.0/usb5/5-4/5-4:1.0/tty/ttyACM0
Mon Apr 13 01:57:54 CEST 2015 DEVPATH=/devices/pci0000:00/0000:00:12.0/usb5/5-4

so it actually get called twice with kill ..

Gentoo does both systemd and non-systemd... i happen to be in the non-systemd camp. 
IMHO systemd is way too much changes not only to the tooling, but also to all stuff surrounding a system for management. And before i leave something rock-solid behin d i would like to be sure that the new stuff is at least as rock-solid.
Personal experience did not support the systemd case. (i did try it unvolutarily on a system where it got installed during an upgrade, that system nog boot with a ton of errors roling on the screen and it boot actually slower then it ever did before.. [ no way back there alas ]).
Another one was installing a CentOS 7 VM which worked, but wouldn't reboot anymore after initial install with all kinds of issues surrounding network adapters.

That's why

**Comment by noci2012** [comment]
Thats why eudev is running on this system.

**Comment by MoonbaseOtago** [comment]
On Sun, 12 Apr 2015 17:06:43 noci2012 wrote:

> this is DEVPATH when invoked from udev.
> Mon Apr 13 01:57:54 CEST 2015
> DEVPATH=/devices/pci0000:00/0000:00:12.0/usb5/5-4/5-4:1.0/tty/ttyACM0 Mon
> Apr 13 01:57:54 CEST 2015 DEVPATH=/devices/pci0000:00/0000:00:12.0/usb5/5-4
> 
> so it actually get called twice with kill ..
> 
> Gentoo does both systemd and non-systemd... i happen to be in the
> non-systemd camp.  IMHO systemd is way too much changes not only to the
> tooling, but also to all stuff surrounding a system for management. And
> before i leave something rock-solid behin d i would like to be sure that
> the new stuff is at least as rock-solid. Personal experience did not
> support the systemd case. (i did try it unvolutarily on a system where it
> got installed during an upgrade, that system nog boot with a ton of errors
> roling on the screen and it boot actually slower then it ever did before..
> [ no way back there alas ]). Another one was installing a CentOS 7 VM which
> worked, but wouldn't reboot anymore after initial install with all kinds of
> issues surrounding network adapters.

I don't think this is a systemd vs. non systemd thing - more like a bug in 
udev - the only distros I've found so far that don't have this bug are from 
red hat - there's a bug they've fixed that hasn't found its way upstream (and 
back down into things like eudev)

Everything in udev gets called twice on install and twice on kill .... the 
main problem is that no one writes scripts for device removal and so any udev 
bugs to do with it don't get fixed - I suspect it's a chicken and egg sort of 
thing - people don't write removal scripts because udev is broken, people move 
on, udev doesn't get fixed - I'd go and poke the udev people if I knew what the 
bug fix actually was - but that wouldn't help my customers today, so I've tried 
to code around the problem

What's likely different is that eudev does something different yet again - what 
we need is to debug that ......

first of all - what's the value of $2 - what we're supposed to get (the real 
bug in udev) is that $2 should be the name of the device "ttyACM0" .....not 
something else - the good news is that ttyACM0 is in your DEVPATH - we 
probably need to do something like:

```
            t1="`echo $DEVPATH | grep ttyACM | wc -l"
            if [ "$v" == "1" ]
            then
                    t1="$DEVPATH"
            else
                    t1="`ls -lt /sys/class/tty/ttyACM* | grep $DEVPATH`"
            fi
            t2=`basename "$t1"`
            if [ -e /var/lock/LCK..$t2 ]
            then
                    kill -9 `cat /var/lock/LCK..$t2`
            fi
```

does that work for you?

```
Paul
```

**Comment by noci2012** [comment]
The udev people, @redhat that totally rebuilt udev into systemd. The only udev that is around separate is eudev from the gentoo maintainers.

How about a test like:

```
if [ -z "$DEVPATH" ] 
then
      echo "Missing DEVPATH variable, are you running from udev?"
      exit 1
fi 
```

before using udev...?

**Comment by MoonbaseOtago** [comment]
On Mon, 13 Apr 2015 11:51:44 noci2012 wrote:

> The udev people, @redhat that totally rebuilt udev into systemd. The only
> udev that is around separate is eudevfrom the gentoo maintainers.

(I must admit I've been studiously avoiding the systemd wars, so I didn't 
realise that - I do embedded stuff mostly - I've also been in this biz for so 
long -  way pre linux - that I've seen init systems come and go, change seems 
to be in the nature of the beast)

hmm I need to grab a recent debian to see if the fix comes with systemd - I was 
a gentoo user for 6-7 years but switch a few years back

> How about a test like:
> 
> it [ -z "$DEVPATH" ] 
> then
>       echo "Missing DEVPATH variable, are you running from udev?"
>       exit 1
> fi 
> 
> before using udev...? 

good idea, it will be in my next release 

```
Paul
```

**Comment by noci2012** [comment]
Here are my fixes for the onerng.sh script.
truncate actually needs -s 0 [ --size or --reference  required ] ..
(+ all of the above incoporated) 

```
--- files/onerng.sh     2015-04-10 06:41:59.000000000 +0200
+++ /sbin/onerng.sh     2015-04-14 21:52:50.887809017 +0200
@@ -32,7 +32,7 @@
        i=0
         while [ $i -lt 200 ]   # loop waiting for things to come up
         do
-               truncate "$t"
+               truncate -s 0 "$t"
                # off, produce nothing, flush
                 echo "cmd0" >/dev/$2           # standard noise
                 echo "cmdO" >/dev/$2           # turn it on
@@ -56,7 +56,7 @@
        then
                sleep 0.1
                # read data into temp file
-               truncate "$t"
+               truncate -s 0 "$t"
                dd if=/dev/$2 iflag=fullblock of=$t bs=1 &
                pid=$!
                sleep 0.02
@@ -115,6 +115,11 @@
 #      when something is removed kill the daemon
 #
 if [ "$1" = "kill" ]; then
+       if [ -z "$DEVPATH" ]
+       then
+               echo "Are you running from udev?"
+               exit 1
+       fi
        if [ -e /var/lock/LCK..$2 ]
        then
                kill -9 `cat /var/lock/LCK..$2`
@@ -128,11 +133,16 @@
                #               matches one of the /sys/class/tty/ttyACM* files
                #               if so use that name's lock file to kill rngd
                #
-               t1="`ls -lt /sys/class/tty/ttyACM* | grep $DEVPATH`"
-               t2=`basename "$t1"`
-               if [ -e /var/lock/LCK..$t2 ]
+               tty=${DEVPATH##*/tty/}                  # this is the device name if any.. or full path
+               btty=$( basename $tty )                 # if tty = path, then btty != ttyname & tty != ttyname...
+                if [ "$tty" != "$btty" ]
+               then
+                       t1=$( ls -lt /sys/class/tty/ttyACM* | grep $DEVPATH )
+                       tty=$( basename "$t1" )         # this is empty if ttyACM doesn't exist....
+               fi
+               if [ -e /var/lock/LCK..$tty ]           # empty $tty or not delivers the right result
                then
-                       kill -9 `cat /var/lock/LCK..$t2`
+                       kill -9 `cat /var/lock/LCK..$tty`
                fi
        fi
         exit 0
```

**Comment by MoonbaseOtago** [comment]
On Tue, 14 Apr 2015 13:02:44 noci2012 wrote:

> Here are my fixes for the onerng.sh script.
> truncate actually needs -s 0 [ --size or --reference  required ] ..
> (+ all of the above incoporated) 

thanks - the truncate fixes are already in the 3.1 release (not on the main 
page quite yet), I'll test those changes in my VM collection and schedule them 
for 3.2

```
thanks

Paul
```

**Comment by noci2012** [comment]
Now the onerng_verify.py script...
i had to remove all ord() occurences to make it work.
i do use python3 ...

```
Traceback (most recent call last):
  File "/sbin/onerng_verify.py", line 147, in <module>
    x = ord(c[length-end_off+0])
TypeError: ord() expected string of length 1, but int found
```

**Comment by MoonbaseOtago** [comment]
On Tue, 14 Apr 2015 13:30:39 noci2012 wrote:

> Now the onerng_verify.py script...
> i had to remove all ord() occurences to make it work.
> i do use python3 ...

yes - several people have pointed that out, I'm still not sure why (works on 
all the systems I've tested it on) - I have someone looking at a portable 
solution

```
Paul
```

**Comment by noci2012** [comment]
ouch there is one glitch in the script.., in the kill path the LCK file is not removed.. fixed below.

```
--- files/onerng.sh     2015-04-10 06:41:59.000000000 +0200
+++ /sbin/onerng.sh     2015-04-14 22:35:35.113707023 +0200
@@ -32,7 +32,7 @@
        i=0
         while [ $i -lt 200 ]   # loop waiting for things to come up
         do
-               truncate "$t"
+               truncate -s 0 "$t"
                # off, produce nothing, flush
                 echo "cmd0" >/dev/$2           # standard noise
                 echo "cmdO" >/dev/$2           # turn it on
@@ -56,7 +56,7 @@
        then
                sleep 0.1
                # read data into temp file
-               truncate "$t"
+               truncate -s 0 "$t"
                dd if=/dev/$2 iflag=fullblock of=$t bs=1 &
                pid=$!
                sleep 0.02
@@ -115,9 +115,15 @@
 #      when something is removed kill the daemon
 #
 if [ "$1" = "kill" ]; then
+       if [ -z "$DEVPATH" ]
+       then
+               echo "Are you running from udev?"
+               exit 1
+       fi
        if [ -e /var/lock/LCK..$2 ]
        then
                kill -9 `cat /var/lock/LCK..$2`
+               rm -f -- /var/lock/LCK..$2
        else
                #
                #       some systems have a broken udev, udevd remove
@@ -128,11 +134,17 @@
                #               matches one of the /sys/class/tty/ttyACM* files
                #               if so use that name's lock file to kill rngd
                #
-               t1="`ls -lt /sys/class/tty/ttyACM* | grep $DEVPATH`"
-               t2=`basename "$t1"`
-               if [ -e /var/lock/LCK..$t2 ]
+               tty=${DEVPATH##*/tty/}                  # this is the device name if any.. or full path
+               btty=$( basename $tty )                 # if tty = path, then btty != ttyname & tty != ttyname...
+                if [ "$tty" != "$btty" ]
+               then
+                       t1=$( ls -lt /sys/class/tty/ttyACM* | grep $DEVPATH )
+                       tty=$( basename "$t1" )         # this is empty if ttyACM doesn't exist....
+               fi
+               if [ -e /var/lock/LCK..$tty ]           # empty $tty or not delivers the right result
                then
-                       kill -9 `cat /var/lock/LCK..$t2`
+                       kill -9 `cat /var/lock/LCK..$tty`
+                       rm -f -- /var/lock/LCK..$tty
                fi
        fi
         exit 0
```

**Comment by MoonbaseOtago** [comment]
On Tue, 14 Apr 2015 13:41:01 noci2012 wrote:

> ouch there is one glitch in the script.., in the kill path the LCK file is
> not removed.. fixed below.

yes that's worth doing, though any locking process should check to see if the 
lock is stale and do it itself

```
Paul
```

**Comment by noci2012** [comment]
If onerng.sh kill ttyACM0 is called two times in a row, there is an error message saying:

```
/sbin/onerng.sh: line 125: kill: (32652) - No such process
```

on the 2nd run.

**Comment by MoonbaseOtago** [comment]
On Tue, 14 Apr 2015 13:47:27 noci2012 wrote:

> If onerng.sh kill ttyACM0 is called two times in a row, there is an error
> message saying:
> 
> ```
> /sbin/onerng.sh: line 125: kill: (32652) - No such process
> ```

it's not the killed  process that cleans up a stale lock - a new process 
trying to make a lock reads the pid from it pokes at that pid to see if the 
pid is still valid and is not takes over the lock

(I haven't looked to see if rngd does the right thing - which is why deleting 
the lock is likely a good thing anyway)

```
Paul
```

**Comment by noci2012** [comment]
I tried all the common signals, it stops on several, but never removes the pid file.
so IMHO the script should remove it.

**Comment by noci2012** [comment]
solved in 3.3
